### PR TITLE
[Serve] Enable `serve status` when proxies disabled

### LINKED ISSUE #41723: [Serve] serve status breaks when disabling http proxy
### What happened + What you expected to happen

When running serve status, i get an error caused by an validation error. 

```
Traceback (most recent call last):
  File "/home/container_user/.local/bin/serve", line 8, in <module>
    sys.exit(cli())
  File "/usr/local/lib/python3.10/dist-packages/click/core.py", line 1130, in __call__
    return self.main(*args, **kwargs)
  File "/usr/local/lib/python3.10/dist-packages/click/core.py", line 1055, in main
    rv = self.invoke(ctx)
  File "/usr/local/lib/python3.10/dist-packages/click/core.py", line 1657, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/usr/local/lib/python3.10/dist-packages/click/core.py", line 1404, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/usr/local/lib/python3.10/dist-packages/click/core.py", line 760, in invoke
    return __callback(*args, **kwargs)
  File "/usr/local/lib/python3.10/dist-packages/ray/serve/scripts.py", line 702, in status
    **ServeSubmissionClient(address).get_serve_details()
  File "/usr/local/lib/python3.10/dist-packages/ray/dashboard/modules/serve/sdk.py", line 100, in get_serve_details
    self._raise_error(response)
  File "/usr/local/lib/python3.10/dist-packages/ray/dashboard/modules/dashboard_sdk.py", line 283, in _raise_error
    raise RuntimeError(
RuntimeError: Request failed with status code 503: Failed to get a response from the controller. The GCS may be down, please retry later: ray::ServeController.get_serve_instance_details() (pid=12884, ip=10.244.182.9, actor_id=bea8e193585be32034f33c4001000000, repr=<ray.serve._private.controller.ServeController object at 0x7ff0518c6230>)
  File "/usr/lib/python3.10/concurrent/futures/_base.py", line 451, in result
    return self.__get_result()
  File "/usr/lib/python3.10/concurrent/futures/_base.py", line 403, in __get_result
    raise self._exception
  File "/usr/local/lib/python3.10/dist-packages/ray/serve/_private/controller.py", line 895, in get_serve_instance_details
    return ServeInstanceDetails(
  File "pydantic/main.py", line 341, in pydantic.main.BaseModel.__init__
pydantic.error_wrappers.ValidationError: 1 validation error for ServeInstanceDetails
proxy_location
  value is not a valid enumeration member; permitted: 'Disabled', 'HeadOnly', 'EveryNode' (type=type_error.enum; enum_values=[<ProxyLocation.Disabled: 'Disabled'>, <ProxyLocation.HeadOnly: 'HeadOnly'>, <ProxyLocation.EveryNode: 'EveryNode'>]).
```

### Versions / Dependencies

2.8.0

### Reproduction script

1. start ray `ray start --head`
2. python ray_debug.py
```
from ray import serve
from ray.serve.config import ProxyLocation

@serve.deployment
class Server:
  def dummy(self):
    print("hello")
    return "hello"

app = Server.bind()
serve.start(detached=True, proxy_location=ProxyLocation.Disabled)
serve.run(app)
```
3. run `serve status`

### Issue Severity

Medium: It is a significant difficulty but I can work around it.

### PR DESCRIPTION
**Pull request by shrekris-anyscale**
## Why are these changes needed?

Serve proxies can be disabled by setting `serve.start(..., proxy_location=ProxyLocation.Disabled)`. However, `serve status` currently fails when proxies are disabled.

This change fixes the issue and lets `serve status` succeed even when `proxy_location==ProxyLocation.Disabled`.

## Related issue number

Closes #41723.

## Checks

- [X] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [X] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [X] Unit tests
     - This change adds a unit test to `test_config.py`.

### CODE REVIEW

File: python/ray/serve/config.py [hunk 1]
```diff
@@ -150,17 +150,36 @@ class ProxyLocation(str, Enum):
...
+
+        if proxy_location == ProxyLocation.Disabled:
             return DeploymentMode.NoServer
-        elif v == ProxyLocation.HeadOnly:
-            return DeploymentMode.HeadOnly
-        elif v == ProxyLocation.EveryNode:
-            return DeploymentMode.EveryNode
         else:
-            raise ValueError(f"Unrecognized `ProxyLocation`: {v}.")
+            return DeploymentMode(proxy_location.value)
```
Comment (shrekris-anyscale) [inline python/ray/serve/config.py:166]: This should not change behavior. I refactored this for simplicity's sake.

File: python/ray/serve/_private/controller.py [hunk 2]
```diff
@@ -929,10 +929,15 @@ def get_serve_instance_details(self) -> Dict:
         # Eventually we want to remove route_prefix from DeploymentSchema.
         http_options = HTTPOptionsSchema.parse_obj(http_config.dict(exclude_unset=True))
         grpc_options = gRPCOptionsSchema.parse_obj(grpc_config.dict(exclude_unset=True))
+        proxy_location = (
+            None
+            if http_config.location is None
+            else ProxyLocation._from_deployment_mode(http_config.location)
```
Comment (GeneDer) [inline python/ray/serve/_private/controller.py:935]: Non-blocker: I feel we can just have an early return in `_from_deployment_mode` to return `None` when `None` is passed so we don't need this if else logics here :)

File: python/ray/serve/tests/unit/test_config.py [hunk 3]
```diff
@@ -545,6 +545,31 @@ def test_proxy_location_to_deployment_mode():
...
+        ProxyLocation._from_deployment_mode(DeploymentMode.EveryNode)
+        == ProxyLocation.EveryNode
+    )
+
+    assert ProxyLocation._from_deployment_mode("NoServer") == ProxyLocation.Disabled
+    assert ProxyLocation._from_deployment_mode("HeadOnly") == ProxyLocation.HeadOnly
+    assert ProxyLocation._from_deployment_mode("EveryNode") == ProxyLocation.EveryNode
+
+    with pytest.raises(ValueError):
+        ProxyLocation._from_deployment_mode("Unknown")
```
Comment (GeneDer) [inline python/ray/serve/tests/unit/test_config.py:567]: Side note: this is an existing behavior, but I think even with different casing like `headonly` will not be able to find the enum as well 😛

File: python/ray/serve/_private/controller.py [hunk 4]
```diff
@@ -929,10 +929,15 @@ def get_serve_instance_details(self) -> Dict:
         # Eventually we want to remove route_prefix from DeploymentSchema.
         http_options = HTTPOptionsSchema.parse_obj(http_config.dict(exclude_unset=True))
         grpc_options = gRPCOptionsSchema.parse_obj(grpc_config.dict(exclude_unset=True))
+        proxy_location = (
```
Comment (GeneDer) [inline python/ray/serve/_private/controller.py:932]: Non-blocker: might be good to add a comment to explain the differences between `DeploymentMode` and `ProxyLocation` so future folks know why this is needed and how it fixes the issue.

**Review by GeneDer — APPROVED** [review]
Small non-blocking suggestions, but overall LGTM! Thanks for fixing the issue Shreyas!

File: python/ray/serve/tests/unit/test_config.py (hunk 3 above)
Comment (shrekris-anyscale) [reply python/ray/serve/tests/unit/test_config.py:567]: That's true, but I don't want to change that behavior in this PR. I'll leave it as-is.

File: python/ray/serve/_private/controller.py (hunk 2 above)
Comment (shrekris-anyscale) [reply python/ray/serve/_private/controller.py]: Sounds good, I added an early return.

File: python/ray/serve/_private/controller.py (hunk 4 above)
Comment (shrekris-anyscale) [reply python/ray/serve/_private/controller.py]: I added a docstring to `_from_deployment_mode`.

**Review by zcin — APPROVED** [review]

**Pull request merged**